Document for ExternalNode feature (#3963)
antrea-io/antrea: docs/external-node.md

# ExternalNode

## Table of Contents

<!-- toc -->
- [What is ExternalNode?](#what-is-externalnode)
- [Prerequisites](#prerequisites)
- [The ExternalNode resource](#the-externalnode-resource)
  - [Name and namespace](#name-and-namespace)
  - [Interfaces](#interfaces)
- [Apply Antrea NetworkPolicy to ExternalNode](#apply-antrea-networkpolicy-to-externalnode)
- [Security](#security)
- [Traffic model](#traffic-model)
  - [OVS pair ports](#ovs-pair-ports)
  - [OpenFlow pipeline](#openflow-pipeline)
    - [Non-IP packet](#non-ip-packet)
    - [IP packet](#ip-packet)
- [Limitations](#limitations)
<!-- /toc -->

## What is ExternalNode?

> `ExternalNode` is a CRD API that manages the network connectivity and security on Non Kubernetes Node (like a virtual

Reviewer: that enables Antrea to manage the network connectivity...

Author: done.

machine or a bare-metal server). It supports specifying which network interfaces on the external Node are expected to be

> guarded with Antrea NetworkPolicy rules. The virtual machine or bare-metal server presented as `ExternalNode` can be

Reviewer: s/presented as `ExternalNode`/represented by an `ExternalNode` resource

> either Linux or Windows. "external Node"

Reviewer: strange line break?

Author: updated.

will be used to represent such a virtual machine or a bare-metal server in this document.

> Antrea NetworkPolicy is applied to `ExternalNode` via the `ExternalEntity` resource.

Reviewer 2: Could you wrap lines at a fixed column (80 to be consistent with other docs).

Author: updated.

`antrea-controller` creates an `ExternalEntity` resource for each network interface defined in the `ExternalNode`
resource.

`antrea-agent` is running on the external Node which is described as `ExternalNode`, and it controls network

> connectivity and security by attaching the network interface on OVS. A new OpenFlow pipeline is implemented dedicated

Reviewer: s/by attaching the network interface on OVS/by attaching the network interface(s) to an OVS bridge

Author: done

> connectivity and security by attaching the network interface on OVS. A new OpenFlow pipeline is implemented dedicated

Reviewer: s/is implemented dedicated for ExternalNode feature/has been implemented, dedicated to the ExternalNode feature

Author: done

for ExternalNode feature.

You may be interested in using this capability for the below scenarios:

> - To apply Antrea NetworkPolicy for an external Node.

Reviewer: s/for an external Node/to an external Node

Author: done

> - You want the same security configurations on the external Node for all the Operating Systems.

Reviewer: s/all the/all

This guide demonstrates how to configure `ExternalNode` to achieve the above result.

## Prerequisites

`ExternalNode` is introduced in v1.8 as an alpha feature. A feature gate `ExternalNode`
must be enabled on the `antrea-controller` and `antrea-agent` in the configuration. The configuration

> for `antrea-controller` is modified in the `antrea-config` ConfigMap like the following options for the feature to work:

Reviewer: s/like the following options for the feature to work:/as follows for the feature to work:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: antrea-config-dcfb6k2hkm
  namespace: kube-system
data:
  antrea-controller.conf: |
    featureGates:
      ExternalNode: true
```

Since `antrea-agent` is running on an external Node which is not managed by Kubernetes, a single configuration file is
required to locate on each machine where `antrea-agent`
is running. The following options for the feature are needed:

```yaml
featureGates:
  ExternalNode: true
nodeType: externalNode

> ```

Reviewer: This is not really helpful without additional instructions on how to use the file?

Reviewer 2: I think this needs not to be put in Prerequisites, as our sample config file for VM agent already enable the feature.

Author: Remove antrea-agent part.

> Other options for `antrea-agent` running on an external Node please refer to

Reviewer: For additional configuration options when running `antrea-agent` on an external Node, please refer to...

the [sample configuration file](../build/yamls/externalnode/conf/antrea-agent.conf).

## The ExternalNode resource

A typical `ExternalNode` resource example:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ExternalNode
metadata:
  name: vm1
  namespace: vm-ns
  labels:
    role: db
spec:
  interfaces:
    - ips: [ "172.16.100.3" ]
      name: ""
```

> ### Name and namespace

Reviewer 2: "namespace" -> "Namespace"

Author: done

- CIDR with an `ipBlock`
- fqdn address in an egress rule

Following actions are supported when applying Antrea NetworkPolicy to an external Node:

- Allow
- Drop
- Reject

Following is an example to apply Antrea NetworkPolicy to the `ExternalNode` objects labeled with `role=db` to reject the
SSH connections from IP "172.16.100.5" or from other external Nodes labeled with `role=front`:

```yaml
kind: NetworkPolicy
metadata:
  name: anp1
  namespace: vm-ns
spec:
  priority: 9000.0
  appliedTo:
    - externalEntitySelector:
        matchLabels:
          role: db
  egress:
    - action: Reject
      ports:
        - protocol: TCP
          port: 22
      to:
        - externalEntitySelector:
            matchLabels:
              role: front
        - ipBlock:
            cidr: 172.16.100.5/32
```

`antrea-agent` running on the target external Node sets up a connection with
`antrea-controller` using the `ExternalNode` name as an option to receive the internal Antrea NetworkPolicy objects.
More details please refer to the
[installation guide](vm-installation.md) for `antrea-agent` on an external Node.

`antrea-agent` realizes the Antrea NetworkPolicy on OVS using OpenFlow entries.

> ## Security

Reviewer 2: Should we name it "RBAC for antrea-agent"?

Author: renamed.

An external Node is thought as an untrusted point on the network. To follow the security principle that least
permissions are given, the RBAC for `antrea-agent`
running on `ExternalNode` feature is modified:

- Only `get`, `list` and `watch` permissions are given on resource `ExternalNode`
- Only `update` permission is given on resource `antreaagentinfos`, and `create`
  permission is moved to `antrea-controller`
- for more details please refer to [vm-agent.yaml](../build/yamls/externalnode/vm-agent.yaml)

`antrea-agent` reports its status by updating the `antreaagentinfo` resource which is configured with the same name as
the `ExternalNode`. `antrea-controller`
creates an `antreaagentinfo` object for each new `ExternalNode`, and then `antrea-agent`
updates it every minute with its latest status. `antreaagentinfo` is deleted by
`antrea-controller` when the `ExternalNode` is removed.

## Traffic model

> ### OVS pair ports

Reviewer 2: Probably name it "VM network interface configuration" or "VM network configuration" and make it the top level.

I am assuming all contents can be interested by regular users. If you feel there are design details users do not care we should separate them out.

Author: Renamed as "VM network configuration", and move as  a separate section.

`antrea-agent` uses the interface IPs or name to find the network interface on the external Node, and then attaches it
on the OVS bridge. The network interface is attached on OVS as uplink, and a new OVS internal Port is added to take over
the uplink interface's IP/MAC and routing configurations. On Windows, the DNS configurations are also moved to the OVS
internal port from uplink. Before attaching the uplink to OVS, the network interface is renamed with a suffix
"~", and OVS internal port is configured with the original name of the uplink. As a result, IP/MAC/routing entries are
seen on a network interface configuring with the same name on the external Node.

The outbound traffic sent from the external Node enters OVS from the internal port, and finally output from the uplink,
and the inbound traffic enters OVS from the uplink and output to the internal port. The IP packet is processed by the
OpenFlow pipeline, and the non-IP packet is forwarded directly.

Following is a diagram for the traffic model on an external Node.
![Traffic On ExternalNode](assets/traffic_external_node.svg)

> ### OpenFlow pipeline

Reviewer 2: This section is not interested by most users. I think we should move it to a separate section.

Author: updated.

packets. Non-IP packet is forwarded between the pair ports directly, e.g., a non-IP packet entering OVS from the uplink
interface is output to the paired internal port, and a packet from the internal port is output to the uplink.

#### IP packet

One OpenFlow pipeline is set on an external Node to process the IP packet. Antrea NetworkPolicy is the major function in
this new pipeline, and the OpenFlow tables used are similar to the container case. **No L3 routing is provided on an
external Node**, and a simple L2 forwarding policy is implemented. OVS connection track is used to assist the
NetworkPolicy function, in a result only the first packet is validated by the OpenFlow entries, and the sequencing
packets in an accepted connection are allowed directly .

**Egress/Ingress Tables**

Table `XgressSecurityClassifierTable` is installed in both `stageEgressSecurity` and
`stageIngressSecurity`, which is used to install the OpenFlow entries to ensure a specific IP traffic to bypass Antrea
NetworkPolicy rules, e.g.,the SSH connection from a specific host to the external Node. These rules can be set with the
item

> `policyBypassRules` in the agent configuration. Following is a configuration example,

Reviewer 2: We should explain this in a separate "user" section, as it is really what users would care. We should not hide it in openflow design.

Reviewer 2: Or considering a sub-section under "Apply Antrea NetworkPolicy to ExternalNode"

Author: Move it as a sub-section under "Apply Antrea NetworkPolicy to ExternalNode"

# ExternalNode

## Table of Contents

<!-- toc -->
- [What is ExternalNode?](#what-is-externalnode)
- [Prerequisites](#prerequisites)
- [The ExternalNode resource](#the-externalnode-resource)
  - [Name and Namespace](#name-and-namespace)
  - [Interfaces](#interfaces)
- [Apply Antrea NetworkPolicy to ExternalNode](#apply-antrea-networkpolicy-to-externalnode)
  - [Antrea NetworkPolicy configuration](#antrea-networkpolicy-configuration)
  - [Bypass Antrea NetworkPolicy](#bypass-antrea-networkpolicy)
- [RBAC for antrea-agent](#rbac-for-antrea-agent)
- [VM network configuration](#vm-network-configuration)
- [OpenFlow pipeline](#openflow-pipeline)
    - [Non-IP packet](#non-ip-packet)
    - [IP packet](#ip-packet)
- [Install Antrea Agent on VM](#install-antrea-agent-on-vm)
  - [Prerequisites on Kubernetes cluster](#prerequisites-on-kubernetes-cluster)
  - [Installation on Linux VM](#installation-on-linux-vm)
    - [Prerequisites](#prerequisites-1)
    - [Installation](#installation)
  - [Installation on Windows VM](#installation-on-windows-vm)
    - [Prerequisites](#prerequisites-2)
    - [Installation](#installation-1)
- [Limitations](#limitations)
<!-- /toc -->

> # What is ExternalNode?

Reviewer: IMO the heading levels are wrong (off by one level)
since you use `# ExternalNode` as the title for the document, all section titles should use a lower level (e.g., `## What is ExternalNode?`)

Author: Adjust in the document.

`ExternalNode` is a CRD API that enables Antrea to manage the network connectivity
and security on a Non Kubernetes Node (like a virtual machine or a bare-metal
server). It supports specifying which network interfaces on the external Node

> are expected to be guarded with Antrea NetworkPolicy rules. The virtual machine

Reviewer: nit: I'd suggest "protected" instead of "guarded" but feel free to ignore if you prefer the current version

Author: update as suggested.

or bare-metal server represented by an `ExternalNode` resource can be either
Linux or Windows. "external Node" will be used to represent such a virtual
machine or a bare-metal server in this document.

> Antrea NetworkPolicy is applied to an external Node via the `ExternalEntity`

Reviewer: ```suggestion
Antrea NetworkPolicies are applied to an external Node by leveraging the `ExternalEntity`
```

resource. `antrea-controller` creates an `ExternalEntity` resource for each
network interface defined in the `ExternalNode` resource.

`antrea-agent` is running on the external Node, and it controls network
connectivity and security by attaching the network interface(s) to an OVS bridge.
A new OpenFlow pipeline has been implemented, dedicated to the ExternalNode
feature.

You may be interested in using this capability for the below scenarios:

- To apply Antrea NetworkPolicy to an external Node.
- You want the same security configurations on the external Node for all
  Operating Systems.

This guide demonstrates how to configure `ExternalNode` to achieve the above
result.

# Prerequisites

> `ExternalNode` is introduced in v1.8 as an alpha feature. A feature gate

Reviewer: The `ExternalNode` feature gate must be enabled...

> `ExternalNode` must be enabled on the `antrea-controller` and `antrea-agent`in

Reviewer: must be enabled in the `antrea-controller` and `antrea-agent` configuration.

the configuration. The configuration for `antrea-controller` is modified in the
`antrea-config` ConfigMap as follows for the feature to work:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: antrea-config-dcfb6k2hkm
  namespace: kube-system
data:
  antrea-controller.conf: |
    featureGates:
      ExternalNode: true
```

Service `antrea` is created for `antrea-controller` to accept connections from
`antrea-agent`, and the type is `ClusterIP` by default. Since an external Node
is running outside the Kubernetes cluster, the `ClusterIP` address is not
accessible. So a `NodePort` or `LoadBalancer` typed Service for `antrea` is

> required.

Reviewer: for this paragraph:

> The `antrea-controller` implements the `antrea` Service, which accepts connections from each `antrea-agent` and is an important part of the NetworkPolicy implementation. By default, the `antrea` Service has type `ClusterIP`. Because external Nodes run outside of the Kubernetes cluster, they cannot directly access the `ClusterIP` address. Therefore, the `antrea` Service needs to become externally-accessible, by changing its type to `NodePort` or `LoadBalancer`.

Author: update as suggested.

Since `antrea-agent` is running on an external Node which is not managed by
Kubernetes, a single configuration file is required to locate on each machine

> where `antrea-agent` is running. 

Reviewer: for the second part of the sentence:

>, a configuration file needs to be present on each machine where the `antrea-agent` is running, and the path to this file will be provided to the `antrea-agent` as a command-line argument.

Author: updated as suggested

> For the detail configuration options when running `antrea-agent`on an external

Reviewer: missing space between "`antrea-agent`" and "on"

Author: added.

Node, please refer to the [sample configuration file](../build/yamls/externalnode/conf/antrea-agent.conf).

> 

Reviewer: maybe we should add a sentence here like this one:

```
A further [section](#install-antrea-agent-on-vm) will provide detailed steps for running the `antrea-agent` on a VM.
```

@jianjuns what do you think?

Author: Added.

Author: updated.

# The ExternalNode resource

A typical `ExternalNode` resource example:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ExternalNode
metadata:
  name: vm1
  namespace: vm-ns
  labels:
    role: db
spec:
  interfaces:
    - ips: [ "172.16.100.3" ]
      name: ""
```

## Name and Namespace

The `name` field in an `ExternalNode` is unique, which is the key to tell
`antrea-agent` where it is running on. It is also used as the key in the
connection from `antrea-agent` to `antrea-controller`. The `ExternalNode` name
is given to `antrea-agent` via an environment variable `NODE_NAME`, and
`antrea-agent` will use the hostname to find the `ExternalNode` resource if it
is not set.

> `ExternalNode` resource is working in `namespace` scope. The `namespace` is

Reviewer: s/is working in/is defined at the

> given to `antrea-agent` with option `externalNodeNamespace` in [antrea-agent.conf](../build/yamls/externalnode/conf/antrea-agent.conf).

Reviewer: The `namespace` name is provided to the `antrea-agent`

BTW usually write "Namespace", not `namespace`, in our documentation

Author: Updated in this document.

```yaml
externalNodeNamespace: vm-ns
```

## Interfaces

The `interfaces` field specifies the list of the network interfaces expected to
be guarded by Antrea NetworkPolicy. One interface at least is required. The
`name` or `ips` can be used to identify the target interface. **The field `ips`
must be provided in the CRD**, but `name` is optional. Multiple IPs on a single
interface is supported.

`antrea-controller` creates an `ExternalEntity` for each interface whenever it

> finds an `ExternalNode is created. The created `ExternalEntity` has the

Reviewer: the markdown formatting is wrong here

suggestion:
> whenever an `ExternalNode` is created.

Author: updated.

following characteristics:

> - It is configured within the same namespace as the `ExternalNode`.

Reviewer: s/namespace/Namespace

- The `name` is generated according to the following principles:
    - Use the `ExternalNode` name directly, if there is only one interface, and
      interface name is not specified.
    - Use the format `$ExternalNode.name-$hash($interface.name)[:5]` for other
      cases.
- The `externalNode` field is set with the `ExternalNode` name.
- The `owner` is referring to the `ExternalNode` resource.
- All labels added on `ExternalNode` are copied to the `ExternalEntity`.
- One `Endpoint` is created per IP in the interface, and every Endpoint's name
  uses the interface name if it is set.

> A sample of the `ExternalEntity` created for the above `ExternalNode` interface
> is,

Reviewer: ```suggestion
The `ExternalEntity` resource created for the above `ExternalNode` interface would look like this:
```

```yaml
apiVersion: crd.antrea.io/v1alpha2
kind: ExternalEntity
metadata:
  labels:
    role: db
  name: vm1
  namespace: vm-ns
  ownerReferences:
    - apiVersion: v1alpha1
      kind: ExternalNode
      name: vm1
      uid: 99b09671-72da-4c64-be93-17185e9781a5
  resourceVersion: "5513"
  uid: 5f360f32-7806-4d2d-9f36-80ce7db8de10
spec:
  endpoints:
    - ip: 172.16.100.3
  externalNode: vm1
```

> Note: **Only one interface is supported for Antrea v1.8**.

Reviewer: this should be mentioned earlier, when describing the `ExternalNode` resource, and not after introducing `ExternalEntity`

# Apply Antrea NetworkPolicy to ExternalNode

## Antrea NetworkPolicy configuration

> Antrea NetworkPolicy applies to an `ExternalNode` by setting the `appliedTo`

Reviewer: An Antrea NetworkPolicy is applied to an `ExternalNode` by providing an `externalEntitySelector` in the `appliedTo` field.

> field with a `externalEntitySelector`. The `ExternalNode` object is converted

Reviewer: is automatically converted

(good to emphasize this info)

Author: done.

to `ExternalEntities`, which are used by `antrea-controller` to calculate the
NetworkPolicy span, and `antrea-agent` running on the `ExternalNode` receives
the internal Antrea NetworkPolicy objects.

> Following types of remote addresses are supported when applying Antrea

Reviewer: Following types of network peers are supported when applying an Antrea...

Author: done

NetworkPolicy to an external Node:

> - ExternalNode with a `externalEntitySelector`

Reviewer: s/a/an

- CIDR with an `ipBlock`
- fqdn address in an egress rule

Following actions are supported when applying Antrea NetworkPolicy to an

> external Node:

Reviewer: when applying an Antrea NetworkPolicy

Author: done.

- Allow
- Drop
- Reject

> Following is an example to apply Antrea NetworkPolicy to the `ExternalNode`

Reviewer: an example of applying an

Author: done

> objects labeled with `role=db` to reject the SSH connections from IP

Reviewer: reject SSH connections from

Author: done.

is deleted by `antrea-controller` when the `ExternalNode` is removed.

# VM network configuration

`antrea-agent` uses the interface IPs or name to find the network interface on
the external Node, and then attaches it on the OVS bridge. The network interface
is attached on OVS as uplink, and a new OVS internal Port is added to take over
the uplink interface's IP/MAC and routing configurations. On Windows, the DNS
configurations are also moved to the OVS internal port from uplink. Before
attaching the uplink to OVS, the network interface is renamed with a suffix
"~", and OVS internal port is configured with the original name of the uplink.
As a result, IP/MAC/routing entries are seen on a network interface configuring
with the same name on the external Node.

The outbound traffic sent from the external Node enters OVS from the internal
port, and finally output from the uplink, and the inbound traffic enters OVS
from the uplink and output to the internal port. The IP packet is processed by
the OpenFlow pipeline, and the non-IP packet is forwarded directly.

Following is a diagram for the traffic model on an external Node.
![Traffic On ExternalNode](assets/traffic_external_node.svg)

> # OpenFlow pipeline

Reviewer: maybe make this section the last one in the document

Author: Do you mean OpenFlow pipeline is put after `Limitation`?

   `antrea-agent` binary from the `bin` directory to the Linux VM.

```bash
make docker-bin
```

2. The `antrea-agent.conf` file specifies agent configuration parameters. Copy
   the [agent configuration file](../build/yamls/externalnode/conf/antrea-agent.conf)
   to the VM and edit the `antrea-agent.conf` file to set `clientConnection`,
   `antreaClientConnection` and `externalNodeNamespace` with the correct values.
   Copy `antrea-agent.antrea.kubeconfig` and `antrea-agent.kubeconfig` files to
   the VM, that were generated in the step 4 and step 5 of
   [Prerequisites on Kubernetes cluster](#prerequisites-on-kubernetes-cluster).

```bash
AGENT_NAMESPACE="vm-ns"
AGENT_CONF_PATH="/etc/antrea"
mkdir -p $AGENT_CONF_PATH
# Copy antrea-agent kubeconfig files
cp ./antrea-agent.kubeconfig $AGENT_CONF_PATH
cp ./antrea-agent.antrea.kubeconfig $AGENT_CONF_PATH
# Update clientConnection and antreaClientConnection
sed -i "s|kubeconfig: |kubeconfig: $AGENT_CONF_PATH/|g" antrea-agent.conf
sed -i "s|#externalNodeNamespace: default|externalNodeNamespace: $AGENT_NAMESPACE|g" antrea-agent.conf
# Copy antrea-agent configuration file
cp ./antrea-agent.conf $AGENT_CONF_PATH
```

> 3. Create `antrea-agent` service. Below is a sample snippet to start

Reviewer: I don't see the `NODE_NAME` variable anywhere? I know you can derive it from the host name, but I think it would be better to be explicit in these instructions.

IMO, we should also show the actual `ExternalNode` resource in the previous section

Author: > I don't see the NODE_NAME variable anywhere? I know you can derive it from the host name, but I think it would be better to be explicit in these instructions.

We mentioned that `NODE_NAME` variable setting in the section `The ExternalNode resource->Name and Namespace`. Here do you mean we should add this in the installation part?

> IMO, we should also show the actual ExternalNode resource in the previous section
Here do you mean we should show the result of command "kubectl get externalNode" in section `The ExternalNode resource`?

@antoninbas

> 

Reviewer 2: Remove an empty line.

Author: Removed

> # ExternalNode

Reviewer 2: Does "External Node" look better?

Author: updated.

    - [Installation steps on Linux VM](#installation-steps-on-linux-vm)
  - [Installation on Windows VM](#installation-on-windows-vm)
    - [Prerequisites on Windows VM](#prerequisites-on-windows-vm)
    - [Installation steps on Windows VM](#installation-steps-on-windows-vm)
- [VM network configuration](#vm-network-configuration)
- [RBAC for antrea-agent](#rbac-for-antrea-agent)
- [Apply Antrea NetworkPolicy to ExternalNode](#apply-antrea-networkpolicy-to-externalnode)
  - [Antrea NetworkPolicy configuration](#antrea-networkpolicy-configuration)
  - [Bypass Antrea NetworkPolicy](#bypass-antrea-networkpolicy)
- [OpenFlow pipeline](#openflow-pipeline)
  - [Non-IP packet](#non-ip-packet)
  - [IP packet](#ip-packet)
- [Limitations](#limitations)
<!-- /toc -->

## What is ExternalNode?

`ExternalNode` is a CRD API that enables Antrea to manage the network connectivity
and security on a Non-Kubernetes Node (like a virtual machine or a bare-metal
server). It supports specifying which network interfaces on the external Node
are expected to be protected with Antrea NetworkPolicy rules. The virtual machine
or bare-metal server represented by an `ExternalNode` resource can be either

> Linux or Windows. An "external Node" will be used to represent such a virtual

Reviewer: Linux or Windows. "external Node" will be used to designate such a virtual machine or bare-metal server in the rest of this document.

machine or a bare-metal server in this document.

Antrea NetworkPolicies are applied to an external Node by leveraging the
`ExternalEntity` resource. `antrea-controller` creates an `ExternalEntity`

> resource for each network interface defined in the `ExternalNode` resource.

Reviewer: s/defined/specified

`antrea-agent` is running on the external Node, and it controls network
connectivity and security by attaching the network interface(s) to an OVS bridge.

> A new OpenFlow pipeline has been implemented, dedicated to the ExternalNode

Reviewer: nit: maybe add a markdown link to `new OpenFlow pipeline` and point it to the `OpenFlow pipeline` section?

cannot directly access the `ClusterIP` address. Therefore, the `antrea` Service
needs to become externally-accessible, by changing its type to `NodePort` or
`LoadBalancer`.

Since `antrea-agent` is running on an external Node which is not managed by
Kubernetes, a configuration file needs to be present on each machine where the
`antrea-agent` is running, and the path to this file will be provided to the
`antrea-agent` as a command-line argument. Refer to the [sample configuration](../build/yamls/externalnode/conf/antrea-agent.conf)
to learn the`antrea-agent` configuration options when running on an external Node.

A further [section](#install-antrea-agent-on-vm) will provide detailed steps
for running the `antrea-agent` on a VM.

## The ExternalNode resource

An example `ExternalNode` resource:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ExternalNode
metadata:
  name: vm1
  namespace: vm-ns
  labels:
    role: db
spec:
  interfaces:
    - ips: [ "172.16.100.3" ]
      name: ""
```

Note: **Only one interface is supported for Antrea v1.8**.

### Name and Namespace

The `name` field in an `ExternalNode` uniquely identifies an external Node.

> The `ExternalNode` name is given to `antrea-agent` via an environment variable

Reviewer: s/given/provided

> `NODE_NAME`, and `antrea-agent` will use the hostname to find the `ExternalNode`

Reviewer 2: and -> otherwise

Author: done

> resource if it is not set.

Reviewer 2: it -> `NODE_NAME`

Author: updated.

`ExternalNode` resource is `Namespace` scoped. The `Namespace` is provided to
`antrea-agent` with option `externalNodeNamespace` in
[antrea-agent.conf](../build/yamls/externalnode/conf/antrea-agent.conf).

```yaml
externalNodeNamespace: vm-ns
```

### Interfaces

The `interfaces` field specifies the list of the network interfaces expected to
be guarded by Antrea NetworkPolicy. At least one interface is required. Interface
`name` or `ips` is used to identify the target interface. **The field `ips`
must be provided in the CRD**, but `name` is optional. Multiple IPs on a single
interface is supported. In the case that multiple `interfaces` are configured,

> `name` must be specified in every `interface`.

Reviewer: s/in every/for every

`antrea-controller` creates an `ExternalEntity` for each interface whenever an

> an `ExternalNode` is created. The created `ExternalEntity` has the following

Reviewer: typo: there is an extra "an"

- The `name` is generated according to the following principles:
  - Use the `ExternalNode` name directly, if there is only one interface, and
    interface name is not specified.
  - Use the format `$ExternalNode.name-$hash($interface.name)[:5]` for other
    cases.
- The `externalNode` field is set with the `ExternalNode` name.
- The `owner` is referring to the `ExternalNode` resource.
- All labels added on `ExternalNode` are copied to the `ExternalEntity`.
- Each IP address of the interface is added as an endpoint in the `endpoints`
  list, and the interface name is used as the endpoint name if it is set.

The `ExternalEntity` resource created for the above `ExternalNode` interface
would look like this:

```yaml
apiVersion: crd.antrea.io/v1alpha2
kind: ExternalEntity
metadata:
  labels:
    role: db
  name: vm1
  namespace: vm-ns
  ownerReferences:
    - apiVersion: v1alpha1
      kind: ExternalNode
      name: vm1
      uid: 99b09671-72da-4c64-be93-17185e9781a5
  resourceVersion: "5513"
  uid: 5f360f32-7806-4d2d-9f36-80ce7db8de10
spec:
  endpoints:
    - ip: 172.16.100.3
  externalNode: vm1
```

## Install Antrea Agent on VM

### Prerequisites on Kubernetes cluster

> 1. Enable `ExternalNode` feature on the `antrea-controller`.

Reviewer: ... and expose the `antrea` Service externally (e.g., as a NodePort Service)

2. Create a NameSpace for `antrea-agent`. This document will use `vm-ns` as an

>    example NameSpace for illustration.

Reviewer: s/NameSpace/Namespace (twice)

   ```bash
   kubectl create ns vm-ns
   ```

3. Create a ServiceAccount, ClusterRole and ClusterRoleBinding for

>    `antrea-agent` as shown below. If you use a different Namespace other than

Reviewer: remove "different" (it is implied by "other")

   sed -i "s|#externalNodeNamespace: default|externalNodeNamespace: $AGENT_NAMESPACE|g" antrea-agent.conf
   # Copy antrea-agent configuration file
   cp ./antrea-agent.conf $AGENT_CONF_PATH
   ```

3. Create `antrea-agent` service. Below is a sample snippet to start
   `antrea-agent` as a service on Ubuntu 18.04 or later:

   ```bash
   AGENT_BIN_PATH="/usr/sbin"
   AGENT_LOG_PATH="/var/log/antrea"
   mkdir -p $AGENT_BIN_PATH
   mkdir -p $AGENT_LOG_PATH
   cat << EOF > /etc/systemd/system/antrea-agent.service
   Description="antrea-agent as a systemd service"
   After=network.target
   [Service]
   ExecStart=$AGENT_BIN_PATH/antrea-agent \
   --config=$AGENT_CONF_PATH/antrea-agent.conf \
   --logtostderr=false \
   --log_file=$AGENT_LOG_PATH/antrea-agent.log
   Restart=on-failure
   [Install]
   WantedBy=multi-user.target
   EOF
   
   sudo systemctl daemon-reload
   sudo systemctl enable antrea-agent
   sudo systemctl start antrea-agent
   ```

4. Configure environment variable `NODE_NAME` if the VM's hostname is different
   from the name defined in the `ExternalNode` resource.

   ```bash
   export NODE_NAME="vm1"

>    ```

Reviewer: does this step come a bit late (why after 3.?). How do I set the env variable when I run antrea-agent as a systemd service?

   files to the VM. Please refer to the step 2 of [Installation on Linux VM](#installation-steps-on-linux-vm)
   section for more information.

   ```powershell
   $WIN_AGENT_CONF_PATH="C:\antrea-agent\conf"
   New-Item -ItemType Directory -Force -Path $WIN_AGENT_CONF_PATH
   # Copy antrea-agent kubeconfig files
   Copy-Item .\antrea-agent.kubeconfig $WIN_AGENT_CONF_PATH
   Copy-Item .\antrea-agent.antrea.kubeconfig $WIN_AGENT_CONF_PATH
   # Copy antrea-agent configuration file
   Copy-Item .\antrea-agent.conf $WIN_AGENT_CONF_PATH
   ```

3. Create `antrea-agent` service using nssm. Below is a sample snippet to start
   `antrea-agent` as a service:

   ```powershell
   $WIN_AGENT_BIN_PATH="C:\antrea-agent"
   $WIN_AGENT_LOG_PATH="C:\antrea-agent\logs"
   New-Item -ItemType Directory -Force -Path $WIN_AGENT_BIN_PATH
   New-Item -ItemType Directory -Force -Path $WIN_AGENT_LOG_PATH
   Copy-Item .\antrea-agent.exe $WIN_AGENT_BIN_PATH
   nssm.exe install antrea-agent $WIN_AGENT_BIN_PATH\antrea-agent.exe --config $WIN_AGENT_CONF_PATH\antrea-agent.conf --log_file $WIN_AGENT_LOG_PATH\antrea-agent.log --logtostderr=false
   nssm.exe start antrea-agent
   ```

> 4. Configure environment variable `NODE_NAME` if the VM's hostname is different

Reviewer: same question as above for Linux

   ```

## VM network configuration

`antrea-agent` uses the interface IPs or name to find the network interface on
the external Node, and then attaches it to the OVS bridge. The network interface
is attached to OVS as uplink, and a new OVS internal Port is created to take over
the uplink interface's IP/MAC and routing configurations. On Windows, the DNS
configurations are also moved to the OVS internal port from uplink. Before
attaching the uplink to OVS, the network interface is renamed with a suffix
"~", and OVS internal port is configured with the original name of the uplink.
As a result, IP/MAC/routing entries are seen on a network interface configuring
with the same name on the external Node.

The outbound traffic sent from the external Node enters OVS from the internal
port, and finally output from the uplink, and the inbound traffic enters OVS
from the uplink and output to the internal port. The IP packet is processed by
the OpenFlow pipeline, and the non-IP packet is forwarded directly.

The following diagram depicts the OVS bridge and traffic forwarding on an
external Node:

> ![Traffic On ExternalNode](assets/traffic_external_node.svg)

Reviewer: nit: I notice there is a *lot* of whitespace around this SVG diagram. Maybe you could crop some of it?

Author: updated.

## RBAC for antrea-agent

An external Node is regarded as an untrusted entity on the network. To follow
the principle of the least privilege is given, the RBAC configuration for

> `antrea-agent` running on `ExternalNode` feature is as follows:

Reviewer: > To follow the least privilege principle, the RBAC configuration for the `antrea-agent` running on an external Node is as follows:

- Only `get`, `list` and `watch` permissions are given on resource `ExternalNode`
- Only `update` permission is given on resource `antreaagentinfos`, and `create`
  permission is moved to `antrea-controller`

> - for more details please refer to [vm-agent-rbac.yml](/build/yamls/externalnode/vm-agent-rbac.yml)

Reviewer: don't make this a bullet point and capitalize "For"

`antrea-agent` reports its status by updating the `antreaagentinfo` resource
which is created with the same name as the `ExternalNode`. `antrea-controller`
creates an `antreaagentinfo` resource for each new `ExternalNode`, and then
`antrea-agent` updates it every minute with its latest status. `antreaagentinfo`
is deleted by `antrea-controller` when the `ExternalNode` is deleted.

## Apply Antrea NetworkPolicy to ExternalNode

### Antrea NetworkPolicy configuration

An Antrea NetworkPolicy is applied to an `ExternalNode` by providing an
`externalEntitySelector` in the `appliedTo` field. **The `ExternalEntity`
resource is automatically created for each interface of an `ExternalNode`**,

> which are used by `antrea-controller` to calculate the NetworkPolicies, and

Reviewer: sentence not grammatically correct

> The `ExternalEntity` resource is automatically created for each interface of an `ExternalNode`. `ExternalEntity` resources are used by `antrea-controller` to process the NetworkPolicies, and each `antrea-agent` (including those running on external Nodes) receives the appropriate internal Antrea NetworkPolicy objects.

`antrea-agent` running on the `ExternalNode` receives the internal Antrea
NetworkPolicy objects.

Following types of (from/to) network peers are supported in an Antrea
NetworkPolicy applied to an external Node:

- ExternalEntities selected by an `externalEntitySelector`
- An `ipBlock`
- A FQDN address in an egress rule

Following actions are supported in an Antrea NetworkPolicy applied to an
external Node:

- Allow
- Drop
- Reject

Below is an example of applying an Antrea NetworkPolicy to the external Nodes
labeled with `role=db` to reject SSH connections from IP "172.16.100.5" or from
other external Nodes labeled with `role=front`:

```yaml
kind: NetworkPolicy
metadata:
  name: anp1
  namespace: vm-ns
spec:
  priority: 9000.0
  appliedTo:
    - externalEntitySelector:
        matchLabels:
          role: db
  ingress:
    - action: Reject
      ports:
        - protocol: TCP
          port: 22
      from:
        - externalEntitySelector:
            matchLabels:
              role: front
        - ipBlock:
            cidr: 172.16.100.5/32
```

### Bypass Antrea NetworkPolicy

> In some cases, users would desire particular traffic to bypass Antrea

Reviewer: s/would desire particular traffic/may want some particular traffic

NetworkPolicy rules on an external Node, e.g.,the SSH connection from a special
host to the external Node. `policyBypassRules` can be added in the agent
configuration to define traffic that needs to bypass NetworkPolicy enforcement.
Below is a configuration example:

```yaml
policyBypassRules:
  - direction: ingress
    protocol: tcp
    cidr: 1.1.1.1/32
    port: 22
```

The `direction` can be `ingress` or `egress`. The supported protocols include:
`tcp`,`udp`, `icmp` and `ip`. The `cidr` gives the peer address, which is the
destination in an `egress` rule, and the source in an `ingress` rule. For `tcp`
and `udp` protocols, the `port` is required to specify the destination port.

## OpenFlow pipeline

A new OpenFlow pipeline is implemented by `antrea-agent` dedicated for
`ExternalNode` feature.

> ![OVS pipeline](assets/ovs-pipeline-external-node.svg)

Reviewer: also a lot of whitespace at the bottom of this svg diagram

Author: updated.

### Non-IP packet

> `NonIPTable` is a new OpenFlow table introduced only on an external Node,

Reviewer: s/on an external Node/on external Nodes

> which is dedicated for all non-IP packets. A non-IP packet is forwarded between

Reviewer: s/dedicated for/dedicated to

the pair ports directly, e.g., a non-IP packet entering OVS from the uplink
interface is output to the paired internal port, and a packet from the internal
port is output to the uplink.

### IP packet

> A new OpenFlow pipeline is set up on an external Node to process the IP packet.

Reviewer: ```suggestion
A new OpenFlow pipeline is set up on external Nodes to process IP packets.
```

Antrea NetworkPolicy enforcement is the major function in this new pipeline, and
the OpenFlow tables used are similar to the Pod pipeline. **No L3 routing is
provided on an external Node**, and a simple L2 forwarding policy is

> implemented. OVS connection track is used to assist the NetworkPolicy function,

Reviewer: s/connection track/connection tracking

> in a result only the first packet is validated by the OpenFlow entries, and the

Reviewer: s/, in a result/; as a result

> sequencing packets in an accepted connection are allowed directly.

Reviewer: s/sequencing/subsequent

of Antrea OVS pipeline.

- L2 Forwarding Tables

`L2ForwardingCalcTable` is used to calculate the expected output port of an IP
packet. As the pair ports with the internal port and uplink always exist on the
OVS bridge, and both interfaces are configured with the same MAC address, the
match condition of an OpenFlow entry in `L2ForwardingCalcTable` uses the input
port number but not the MAC address of the packet. The flow actions are:

1) set flag `OFPortFoundRegMark`, and
2) set the peer port as the `TargetOFPortField`, and
3) enforce the packet to go to stageIngressSecurity.

Below is an example OpenFlow entry in `L2ForwardingCalcTable`

```yaml
table=L2ForwardingCalc, priority=200,ip,in_port=ens224 actions=load:0x1->NXM_NX_REG0[8],load:0x7->NXM_NX_REG1[],resubmit(,IngressSecurityClassifier)
table=L2ForwardingCalc, priority=200,ip,in_port="ens224~" actions=load:0x1->NXM_NX_REG0[8],load:0x8->NXM_NX_REG1[],resubmit(,IngressSecurityClassifier)
```

## Limitations

This feature currently supports only one interface per `ExternalNode` object,
and `ips` must be set in the interface. The support for multiple network
interfaces will be added in the future.

> `ExternalNode` name must be unique in the `cluster` scope.

Reviewer: `ExternalNode` name must be unique in the `cluster` scope even though it is itself a Namespaced resource.

   the [agent configuration file](../build/yamls/externalnode/conf/antrea-agent.conf)
   to the VM and edit the `antrea-agent.conf` file to set `clientConnection`,
   `antreaClientConnection` and `externalNodeNamespace` with the correct values.
   Copy `antrea-agent.antrea.kubeconfig` and `antrea-agent.kubeconfig` files to
   the VM, that were generated in the step 4 and step 5 of
   [Prerequisites on Kubernetes cluster](#prerequisites-on-kubernetes-cluster).

   ```bash
   AGENT_NAMESPACE="vm-ns"
   AGENT_CONF_PATH="/etc/antrea"
   mkdir -p $AGENT_CONF_PATH
   # Copy antrea-agent kubeconfig files
   cp ./antrea-agent.kubeconfig $AGENT_CONF_PATH
   cp ./antrea-agent.antrea.kubeconfig $AGENT_CONF_PATH
   # Update clientConnection and antreaClientConnection
   sed -i "s|kubeconfig: |kubeconfig: $AGENT_CONF_PATH/|g" antrea-agent.conf
   sed -i "s|#externalNodeNamespace: default|externalNodeNamespace: $AGENT_NAMESPACE|g" antrea-agent.conf
   # Copy antrea-agent configuration file
   cp ./antrea-agent.conf $AGENT_CONF_PATH
   ```

3. Configure environment variable `NODE_NAME` if the VM's hostname is different
   from the name defined in the `ExternalNode` resource.

   ```bash

>    export NODE_NAME="vm1"

Reviewer: again, I am not sure how this can work for a systemd service. It doesn't feel like this environment variable will be available to the antrea-agent process.

Instead I would expect something like this in `/etc/systemd/system/antrea-agent.service`:
```
[Service]
Environment=NODE_NAME=vm1
ExecStart=...
```

But I am not an expert on this so I may be missing something...

Author: Since this part is copied from the original vm-installation.md,  @Anandkumar26 does this setting work for service? If yes, I would update it in this PR.

Author: Move the environment variable setting into the service configuration.

Reviewer 3: We can update the environment variable this way.
```shell
[Service]
Environment="NODE_NAME=$NODE_NAME"
Environment="EXTERNAL_NODE=$EXTERNAL_NODE"
```
https://gitlab.eng.vmware.com/antreaplus/antreacloud/-/blob/master/build/images/agent/linux/install.sh#L527

Author: Variable "EXTERNAL_NODE" is not introduced in the feature branch

`ClusterIP`. Because external Nodes run outside of the Kubernetes cluster, they
cannot directly access the `ClusterIP` address. Therefore, the `antrea` Service
needs to become externally-accessible, by changing its type to `NodePort` or
`LoadBalancer`.

Since `antrea-agent` is running on an external Node which is not managed by
Kubernetes, a configuration file needs to be present on each machine where the
`antrea-agent` is running, and the path to this file will be provided to the
`antrea-agent` as a command-line argument. Refer to the [sample configuration](../build/yamls/externalnode/conf/antrea-agent.conf)
to learn the`antrea-agent` configuration options when running on an external Node.

A further [section](#install-antrea-agent-on-vm) will provide detailed steps
for running the `antrea-agent` on a VM.

## The ExternalNode resource

An example `ExternalNode` resource:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ExternalNode
metadata:
  name: vm1
  namespace: vm-ns
  labels:
    role: db
spec:
  interfaces:
    - ips: [ "172.16.100.3" ]
      name: ""
```

Note: **Only one interface is supported for Antrea v1.8**.

### Name and Namespace

> The `name` field in an `ExternalNode` must be unique, which is the key for

Author: Removed the words.

```

### Interfaces

The `interfaces` field specifies the list of the network interfaces expected to
be guarded by Antrea NetworkPolicy. At least one interface is required. Interface
`name` or `ips` is used to identify the target interface. **The field `ips`
must be provided in the CRD**, but `name` is optional. Multiple IPs on a single
interface is supported. In the case that multiple `interfaces` are configured,
`name` must be specified in every `interface`.

`antrea-controller` creates an `ExternalEntity` for each interface whenever an
an `ExternalNode` is created. The created `ExternalEntity` has the following
characteristics:

- It is configured within the same Namespace as the `ExternalNode`.
- The `name` is generated according to the following principles:
  - Use the `ExternalNode` name directly, if there is only one interface, and
    interface name is not specified.
  - Use the format `$ExternalNode.name-$hash($interface.name)[:5]` for other
    cases.
- The `externalNode` field is set with the `ExternalNode` name.
- The `owner` is referring to the `ExternalNode` resource.
- All labels added on `ExternalNode` are copied to the `ExternalEntity`.

> - Each IP address of the interface is added as an Endpoint in the `endpoints`

Reviewer 2: Should we use "endpoint" instead of "Endpoint"?

Author: updated.

which are used by `antrea-controller` to calculate the NetworkPolicies, and
`antrea-agent` running on the `ExternalNode` receives the internal Antrea
NetworkPolicy objects.

Following types of (from/to) network peers are supported in an Antrea
NetworkPolicy applied to an external Node:

- ExternalEntities selected by an `externalEntitySelector`
- An `ipBlock`
- A FQDN address in an egress rule

Following actions are supported in an Antrea NetworkPolicy applied to an
external Node:

- Allow
- Drop
- Reject

Below is an example of applying an Antrea NetworkPolicy to the external Nodes
labeled with `role=db` to reject SSH connections from IP "172.16.100.5" or from
other external Nodes labeled with `role=front`:

```yaml
kind: NetworkPolicy
metadata:
  name: anp1
  namespace: vm-ns
spec:
  priority: 9000.0
  appliedTo:
    - externalEntitySelector:
        matchLabels:
          role: db
  ingress:
    - action: Reject
      ports:
        - protocol: TCP
          port: 22
      from:
        - externalEntitySelector:
            matchLabels:
              role: front
        - ipBlock:
            cidr: 172.16.100.5/32
```

### Bypass Antrea NetworkPolicy

> In some cases, the special IP traffic is expected to bypass the Antrea

Reviewer 2: "users would desire particular traffic to bypass Antrea NetworkPolicy rules on an external Node"

Author: done.

> NetworkPolicy rules on an external Node, e.g.,the SSH connection from a special

Reviewer 2: connections

special -> specific

Author: done.

> host to the external Node. These rules can be set with the item

Reviewer 2: "`policyBypassRules` can be added in the agent configuration to define traffic that needs to bypass NetworkPolicy enforcement"?

Author: updated.

> `policyBypassRules` in the agent  configuration. Following is a configuration

Reviewer 2: Following -> Below

Author: updated.

> example,

Reviewer 2: , -> :

Author: done

```yaml
policyBypassRules:
  - direction: ingress
    protocol: tcp
    cidr: 1.1.1.1/32
    port: 22
```

> The `direction` in one policyBypassRule supports `ingress` or `egress`. And

Reviewer 2: The `direction` can be `ingress` or `egress`

Remove "And".

Author: done.

> the supported protocols include, `tcp`,`udp`, `icmp` and `ip`. The `cidr`

Reviewer 2: include:

Author: updated.

> gives the peer address, it is the destination in an `egress` rule, and the

Reviewer 2: it is -> which is

Author: updated.

source in an `ingress` rule. For `tcp` and `udp` protocols, the `port` is

> required to describe the destination port in the packet.

Reviewer 2: to specify the destination port

Maybe remove "in the packet"

Author: done.

> The policyBypassRules are processed before Antrea NetworkPolicy rules when

Reviewer 2: I feel we should remove this sentence or move to a proper place in "OpenFlow pipeline"

Author: removed this sentence.

forwarding the packets in OVS datapath, and the disposition of the rules is to
allow the packets.

## OpenFlow pipeline

A new OpenFlow pipeline is implemented by `antrea-agent` dedicated for
`ExternalNode` feature.

![OVS pipeline](assets/ovs-pipeline-external-node.svg)

### Non-IP packet

> `NonIPTable` is a new OpenFlow table introduced only when on an external Node,

Reviewer 2: Remove "when"

Author: done

> which is dedicated for all non-IP packets. Non-IP packet is forwarded between

Reviewer 2: A non-IP

Author: updated.

the pair ports directly, e.g., a non-IP packet entering OVS from the uplink
interface is output to the paired internal port, and a packet from the internal
port is output to the uplink.

### IP packet

> One OpenFlow pipeline is set on an external Node to process the IP packet.

Reviewer 2: "A new OpenFlow pipeline"? Not very sure..

set up

Author: updated.

> Antrea NetworkPolicy is the major function in this new pipeline, and the

Reviewer 2: Antrea NetworkPolicy enforcement

Author: updated.

> OpenFlow tables used are similar to the container case. **No L3 routing is

Reviewer 2: to the Pod pipeline

Author: updated.

provided on an external Node**, and a simple L2 forwarding policy is
implemented. OVS connection track is used to assist the NetworkPolicy function,
in a result only the first packet is validated by the OpenFlow entries, and the
sequencing packets in an accepted connection are allowed directly.

- Egress/Ingress Tables

Table `XgressSecurityClassifierTable` is installed in both `stageEgressSecurity`
and `stageIngressSecurity`, which is used to install the OpenFlow entries for
the [`policyBypassRules`](#bypass-antrea-networkpolicy) in the agent configuration.

> This is an example of the OpenFlow entry for the above configuration,

Reviewer 2: , -> :

Author: done.

```yaml
table=IngressSecurityClassifier, priority=200,ct_state=+new+trk,tcp,nw_src=1.1.1.1,tp_dst=22 actions=resubmit(,IngressMetric)
```

Other OpenFlow tables in `stageEgressSecurity` and `stageIngressSecurity` are

> the same as those installed on Kubernetes worker Node. For more details about

Reviewer 2: a Kubernetes worker Node

Author: done.

these tables, please refer to the general [introduction](design/ovs-pipeline.md)
of Antrea OVS pipeline.

- L2 Forwarding Tables

`L2ForwardingCalcTable` is used to calculate the expected output port of an IP
packet. As the pair ports with the internal port and uplink always exist on the
OVS bridge, and both interfaces are configured with the same MAC address, the

> match condition of the OpenFlow entry in `L2ForwardingCalcTable` is the input

Reviewer 2: an OpenFlow entry?

Or: the OpenFlow entries

Reviewer 2: is -> uses

Author: updated.

> port number but not the MAC address of the packet. The actions include,

Reviewer 2: The flow actions?

include -> are?

, -> :

Author: updated.

1) set flag `OFPortFoundRegMark`, and
2) set the peer port as the `TargetOFPortField`, and
3) enforce the packet to go to stageIngressSecurity.

> Following is an example of the OpenFlow entries in `L2ForwardingCalcTable`

Reviewer 2: Below

Reviewer 2: an example OpenFlow entry

Author: updated.

```yaml
table=L2ForwardingCalc, priority=200,ip,in_port=ens224 actions=load:0x1->NXM_NX_REG0[8],load:0x7->NXM_NX_REG1[],resubmit(,IngressSecurityClassifier)
table=L2ForwardingCalc, priority=200,ip,in_port="ens224~" actions=load:0x1->NXM_NX_REG0[8],load:0x8->NXM_NX_REG1[],resubmit(,IngressSecurityClassifier)
```

## Limitations

This feature currently supports only one interface per `ExternalNode` object,
and `ips` must be set in the interface. The support for multiple network
interfaces will be added in the future.

> `ExternalNode` name is unique in the `cluster` scope.

Reviewer 2: is -> must be

Author: done.

which is configured with the same name as the `ExternalNode`. `antrea-controller`
creates an `antreaagentinfo` object for each new `ExternalNode`, and then
`antrea-agent` updates it every minute with its latest status. `antreaagentinfo`
is deleted by `antrea-controller` when the `ExternalNode` is removed.

## Apply Antrea NetworkPolicy to ExternalNode

### Antrea NetworkPolicy configuration

An Antrea NetworkPolicy is applied to an `ExternalNode` by providing an
`externalEntitySelector` in the `appliedTo` field. **The `ExternalNode` object
is automatically converted to `ExternalEntities`**, which are used by
`antrea-controller` to calculate the NetworkPolicy span, and `antrea-agent`
running on the `ExternalNode` receives the internal Antrea NetworkPolicy
objects.

Following types of network peers are supported when applying an Antrea
NetworkPolicy to an external Node:

- ExternalNode with an `externalEntitySelector`
- CIDR with an `ipBlock`
- fqdn address in an egress rule

Following actions are supported when applying an Antrea NetworkPolicy to an
external Node:

- Allow
- Drop
- Reject

Following is an example of applying an Antrea NetworkPolicy to the `ExternalNode`

> objects labeled with `role=db` to reject SSH connections from IP "172.16.100.5"

Reviewer 4: The description here and the yaml below do not match.

or from other external Nodes labeled with `role=front`:

```yaml
kind: NetworkPolicy
metadata:
  name: anp1
  namespace: vm-ns
spec:
  priority: 9000.0
  appliedTo:
    - externalEntitySelector:
        matchLabels:
          role: db

>   egress:

Reviewer 4: It should be `ingress` and `from` below since we want to reject the `from` traffic in the description.

Author: updated.

```yaml
table=IngressSecurityClassifier, priority=200,ct_state=+new+trk,tcp,nw_src=1.1.1.1,tp_dst=22 actions=resubmit(,IngressMetric)
```

Other OpenFlow tables in `stageEgressSecurity` and `stageIngressSecurity` are
the same as those installed on Kubernetes worker Node. For more details about
these tables, please refer to the general [introduction](design/ovs-pipeline.md)
of Antrea OVS pipeline.

**L2 Forwarding Tables**

`L2ForwardingCalcTable` is used to calculate the expected output port of an IP
packet. As the pair ports with the internal port and uplink always exist on the
OVS bridge, and both interfaces are configured with the same MAC address, the
match condition of the OpenFlow entry in `L2ForwardingCalcTable` is the input
port number but not the MAC address of the packet. The actions include,
1) set flag `OFPortFoundRegMark`, and
2) set the peer port as the `TargetOFPortField`, and
3) enforce the packet to go to stageIngressSecurity.

Following is an example of the OpenFlow entries in `L2ForwardingCalcTable`

```yaml
table=L2ForwardingCalc, priority=200,ip,in_port=ens224 actions=load:0x1->NXM_NX_REG0[8],load:0x7->NXM_NX_REG1[],resubmit(,IngressSecurityClassifier)

> table=L2ForwardingCalc, priority=200,ip,in_port="phy-ens224" actions=load:0x1->NXM_NX_REG0[8],load:0x8->NXM_NX_REG1[],resubmit(,IngressSecurityClassifier)

Reviewer 4: `phy-ens224` should be `ens224~`?

# ExternalNode

## Table of Contents

<!-- toc -->
- [What is ExternalNode?](#what-is-externalnode)
- [Prerequisites](#prerequisites)
- [The ExternalNode resource](#the-externalnode-resource)
  - [Name and Namespace](#name-and-namespace)
  - [Interfaces](#interfaces)
- [Apply Antrea NetworkPolicy to ExternalNode](#apply-antrea-networkpolicy-to-externalnode)
  - [Antrea NetworkPolicy configuration](#antrea-networkpolicy-configuration)
  - [Bypass Antrea NetworkPolicy](#bypass-antrea-networkpolicy)
- [RBAC for antrea-agent](#rbac-for-antrea-agent)
- [VM network configuration](#vm-network-configuration)

> - [Install Antrea Agent on VM](#install-antrea-agent-on-vm)

Reviewer 2: Probably put "VM network configuration" before "RBAC for antrea-agent", as it may be more natural to talk about VM configuration changes after installation.

> - [Install Antrea Agent on VM](#install-antrea-agent-on-vm)

Author: If "VM network configuration" and "RBAC for antrea-agent" are about installation, do you think I should put these two under section "Install Antrea Agnet on VM" ? @jianjuns

> - [Install Antrea Agent on VM](#install-antrea-agent-on-vm)

Author: According to the offline discussion, I would leave the two sections at the same level as "VM network configuration"

  - [Prerequisites on Kubernetes cluster](#prerequisites-on-kubernetes-cluster)
  - [Installation on Linux VM](#installation-on-linux-vm)
    - [Prerequisites](#prerequisites-1)
    - [Installation](#installation)
  - [Installation on Windows VM](#installation-on-windows-vm)
    - [Prerequisites](#prerequisites-2)
    - [Installation](#installation-1)
- [OpenFlow pipeline](#openflow-pipeline)
    - [Non-IP packet](#non-ip-packet)
    - [IP packet](#ip-packet)
- [Limitations](#limitations)
<!-- /toc -->

## What is ExternalNode?

`ExternalNode` is a CRD API that enables Antrea to manage the network connectivity

> and security on a Non Kubernetes Node (like a virtual machine or a bare-metal

Author: updated.

server). It supports specifying which network interfaces on the external Node
are expected to be protected with Antrea NetworkPolicy rules. The virtual machine
or bare-metal server represented by an `ExternalNode` resource can be either

> Linux or Windows. "external Node" will be used to represent such a virtual

Author: I change to `An "external Node"`

`ExternalNode` must be enabled in the `antrea-controller` and `antrea-agent`
configuration. The configuration for `antrea-controller` is modified in the
`antrea-config` ConfigMap as follows for the feature to work:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: antrea-config-dcfb6k2hkm
  namespace: kube-system
data:
  antrea-controller.conf: |
    featureGates:
      ExternalNode: true
```

The `antrea-controller` implements the `antrea` Service, which accepts
connections from each `antrea-agent` and is an important part of the
NetworkPolicy implementation. By default, the `antrea` Service has type
`ClusterIP`. Because external Nodes run outside of the Kubernetes cluster, they
cannot directly access the `ClusterIP` address. Therefore, the `antrea` Service
needs to become externally-accessible, by changing its type to `NodePort` or
`LoadBalancer`.

Since `antrea-agent` is running on an external Node which is not managed by
Kubernetes, a configuration file needs to be present on each machine where the
`antrea-agent` is running, and the path to this file will be provided to the
`antrea-agent` as a command-line argument.

> For the detail configuration options when running `antrea-agent` on an external

Author: updated.

> Refer to the [sample configuration file](../build/yamls/externalnode/conf/antrea-agent.conf)

Author: Since the next sentence is a referrence to the installation part, it is a bit strange to put it before this one. Do you think it is better if I remove the configuration referrence sentence?

> Refer to the [sample configuration file](../build/yamls/externalnode/conf/antrea-agent.conf)

Reviewer 2: Sorry, I did not express clearly. I meant merge this sentence to the previous paragraph (line 88).

> Refer to the [sample configuration file](../build/yamls/externalnode/conf/antrea-agent.conf)

Author: Got it, updated.

connection from `antrea-agent` to `antrea-controller`. The `ExternalNode` name
is given to `antrea-agent` via an environment variable `NODE_NAME`, and
`antrea-agent` will use the hostname to find the `ExternalNode` resource if it
is not set.

`ExternalNode` resource is defined at the `Namespace` scope. The `Namespace` is
provided to `antrea-agent` with option `externalNodeNamespace` in
[antrea-agent.conf](../build/yamls/externalnode/conf/antrea-agent.conf).

```yaml
externalNodeNamespace: vm-ns
```

### Interfaces

The `interfaces` field specifies the list of the network interfaces expected to
be guarded by Antrea NetworkPolicy. One interface at least is required. The
`name` or `ips` can be used to identify the target interface. **The field `ips`
must be provided in the CRD**, but `name` is optional. Multiple IPs on a single
interface is supported.

`antrea-controller` creates an `ExternalEntity` for each interface whenever an
an `ExternalNode` is created. The created `ExternalEntity` has the following
characteristics:

- It is configured within the same Namespace as the `ExternalNode`.
- The `name` is generated according to the following principles:

>   - Use the `ExternalNode` name directly, if there is only one interface, and

Reviewer 2: Do not see you add the sentence.

>   - Use the `ExternalNode` name directly, if there is only one interface, and

Author: I added in the first paragraph in `Interfaces` section:
```
The `interfaces` field specifies the list of the network interfaces expected to
be guarded by Antrea NetworkPolicy. At least one interface is required. Interface
`name` or `ips` is used to identify the target interface. **The field `ips`
must be provided in the CRD**, but `name` is optional. Multiple IPs on a single
interface is supported. In the case that multiple `interfaces` are configured,
`name` must be specified in every `interface`.
```

>   - Use the `ExternalNode` name directly, if there is only one interface, and

Reviewer 2: Ok

the external Node, and then attaches it on the OVS bridge. The network interface
is attached on OVS as uplink, and a new OVS internal Port is added to take over
the uplink interface's IP/MAC and routing configurations. On Windows, the DNS
configurations are also moved to the OVS internal port from uplink. Before
attaching the uplink to OVS, the network interface is renamed with a suffix
"~", and OVS internal port is configured with the original name of the uplink.
As a result, IP/MAC/routing entries are seen on a network interface configuring
with the same name on the external Node.

The outbound traffic sent from the external Node enters OVS from the internal
port, and finally output from the uplink, and the inbound traffic enters OVS
from the uplink and output to the internal port. The IP packet is processed by
the OpenFlow pipeline, and the non-IP packet is forwarded directly.

Following is a diagram for the traffic model on an external Node.
![Traffic On ExternalNode](assets/traffic_external_node.svg)

## RBAC for antrea-agent

An external Node is thought as an untrusted point on the network. To follow the

> security principle that least permissions are given, the RBAC for `antrea-agent`

Author: updated

- Only `get`, `list` and `watch` permissions are given on resource `ExternalNode`
- Only `update` permission is given on resource `antreaagentinfos`, and `create`
  permission is moved to `antrea-controller`
- for more details please refer to [vm-agent-rbac.yaml](/build/yamls/externalnode/vm-agent-rbac.yml)

`antrea-agent` reports its status by updating the `antreaagentinfo` resource
which is configured with the same name as the `ExternalNode`. `antrea-controller`
creates an `antreaagentinfo` object for each new `ExternalNode`, and then
`antrea-agent` updates it every minute with its latest status. `antreaagentinfo`
is deleted by `antrea-controller` when the `ExternalNode` is removed.

## Apply Antrea NetworkPolicy to ExternalNode

### Antrea NetworkPolicy configuration

An Antrea NetworkPolicy is applied to an `ExternalNode` by providing an
`externalEntitySelector` in the `appliedTo` field. **The `ExternalNode` object
is automatically converted to `ExternalEntities`**, which are used by
`antrea-controller` to calculate the NetworkPolicy span, and `antrea-agent`
running on the `ExternalNode` receives the internal Antrea NetworkPolicy
objects.

> Following types of network peers are supported when applying an Antrea

Reviewer 2: Maybe add "(from/to) network peers" to be clearer?

Reviewer 2: in an Antrea NetworkPolicy applied to external Nodes:

NetworkPolicy to an external Node:

> - ExternalNode with an `externalEntitySelector`

Reviewer 2: ExternalEntities selected by an `externalEntitySelector`?

> - CIDR with an `ipBlock`

Reviewer 2: Just say "an `ipBlock`"?

> - fqdn address in an egress rule

Reviewer 2: A FQDN address

> Following actions are supported when applying an Antrea NetworkPolicy to an

Reviewer 2: in an Antrea NetworkPolicy applied to external Nodes:

external Node:

- Allow
- Drop
- Reject

> Following is an example of applying an Antrea NetworkPolicy to the `ExternalNode`

Reviewer 2: Following -> Below

> objects labeled with `role=db` to reject SSH connections from IP "172.16.100.5"

Reviewer 2: `ExternalNode` objects -> `ExternalNodes` or "external Nodes"

> or from other external Nodes labeled with `role=front`:

Reviewer 2: Use the same term here and in line 451; either "external Nodes" or "ExternalNodes".

```yaml
kind: NetworkPolicy
metadata:
  name: anp1
  namespace: vm-ns
spec:
  priority: 9000.0
  appliedTo:
    - externalEntitySelector:
        matchLabels:
          role: db
  ingress:
    - action: Reject
      ports:
        - protocol: TCP
          port: 22
      from:
        - externalEntitySelector:
            matchLabels:
              role: front
        - ipBlock:
            cidr: 172.16.100.5/32
```

> `antrea-agent` running on the target external Node sets up a connection with

Reviewer 2: the -> an

> `antrea-agent` running on the target external Node sets up a connection with

Reviewer 2: I feel we can remove this and next paragraphs, as most of this doc and this section are about user guide, but not implementation, and the added information is not very useful either.

At least merge the two paragraphs if you want to keep them.

`antrea-controller` using the `ExternalNode` name as an option to receive the
internal Antrea NetworkPolicy objects.

> `antrea-agent` realizes the Antrea NetworkPolicy on OVS using OpenFlow entries.

Reviewer 2: NetworkPolicies
